### PR TITLE
speed-up list shrinker

### PR DESCRIPTION
**Pull request by jmid**
An improvement to the built-in list shrinker.

### CODE REVIEW

File: src/core/QCheck.ml [hunk 1]
```diff
@@ -1506,10 +1543,11 @@ module Test = struct
     let hist_size, bucket_size =
       let sample_width = Int64.(sub (of_int max_idx) (of_int min_idx)) in
       if sample_width > Int64.of_int stat_max_lines
-      then 1+stat_max_lines,
```
Comment (c-cube) [inline src/core/QCheck.ml]: does this belong in the other PR?

File: src/core/QCheck.ml [hunk 2]
```diff
@@ -395,8 +395,45 @@ module Shrink = struct
         done
 
   let list ?shrink l yield =
```
Comment (c-cube) [inline src/core/QCheck.ml]: looks good. I'll refactor the style a bit but the idea is good in any case (similar to integer shrinking I think).

File: src/core/QCheck.ml (hunk 1 above)
Comment (jmid) [reply src/core/QCheck.ml]: Hm. That's weird. Indeed it has nothing to do with list shrinking (and was committed a day earlier). I *really* should start using branches... :-)

File: src/core/QCheck.ml (hunk 2 above)
Comment (jmid) [reply src/core/QCheck.ml]: Great. Besides the repetitive testing for empty list `[]` the code should mirror the array-based strategy from earlier, as confirmed by matching the exact number of successful shrink steps between the two.

**Comment by c-cube** [comment]
Thank you! :+1:

**Pull request merged**